sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/plans/AliasAwareOutputExpression.scala

```diff
@@ -0,0 +1,147 @@
+        // occurrence of an aliased child. This is to keep the previous behavior and give precedence
+        // the last Alias during `normalizeExpression()` to avoid any kind of regression.
+        a.toAttribute +=:
+          aliases.getOrElseUpdate(strip(child.canonicalized), mutable.ListBuffer.empty)
+      case _ =>
+    }
+    // Append identity mapping of an attribute to the map if both the attribute and its aliased
+    // version can be found in `outputExpressions`.
+    outputExpressions.foreach {
+      case a: Attribute if aliases.contains(a.canonicalized) => aliases(a.canonicalized) += a
+      case _ =>
+    }
+
+    aliases.partition { case (expr, _) => expr.children.exists(_.exists(aliases.contains)) }
+  }
+
+  protected def hasAlias: Boolean = attrAliasMap.nonEmpty
+
+  /**
+   * Return a set of Expression which normalize the original expression to the aliased.
+   */
+  protected def normalizeExpression(expr: Expression): Seq[Expression] = {
+    val outputSet = AttributeSet(outputExpressions.map(_.toAttribute))
+
+    def f: PartialFunction[Expression, Stream[Expression]] = {
+      // Mapping with aliases
+      case e: Expression if exprAliasMap.contains(e.canonicalized) =>
+        (exprAliasMap(e.canonicalized) :+ e).toStream
+      case e: Expression if attrAliasMap.contains(e.canonicalized) =>
+        attrAliasMap(e.canonicalized).toStream
+
+      // Prune if we encounter an attribute that we can't map and it is not in output set.
+      // This prune will go up to the closest `multiTransformDown()` call and returns `Stream.empty`
+      // there.
+      case a: Attribute if !outputSet.contains(a) => Stream.empty
+
+      // Remove `PartitioningCollection` elements that are expressions and contain an attribute that
+      // can't be mapped and the node's output set doesn't contain the attribute.
+      // To achieve this we need to "restart" `multiTransformDown()` for each expression child and
+      // filter out empty streams due to the above attribute pruning case.
+      // The child streams can be then combined using `generateChildrenSeq()` into one stream as
+      // `multiTransformDown()` would also do (but without filtering empty streams).
+      case p: PartitioningCollection =>
+        val childrenStreams = p.partitionings.map {
+          case e: Expression => e.multiTransformDown(f).asInstanceOf[Stream[Partitioning]]
+          case o => Stream(o)
+        }.filter(_.nonEmpty)
+        generateChildrenSeq(childrenStreams).flatMap {
+          case Nil => None
+          // We might have an expression type partitioning that doesn't need
+          // `PartitioningCollection`
+          case (p: Expression) :: Nil => Some(p)
+          case p :: Nil => Some(PartitioningCollection(Seq(p)))
+          case ps => Some(PartitioningCollection(ps))
+        }
+
+      // Filter `SortOrder` children similarly to `PartitioningCollection` elements
+      case s: SortOrder =>
```

Comment:
Since `SortOrder` is always the root of an expression tree, is it better to handle it here? https://github.com/apache/spark/pull/37525/files#diff-2d06454bd3d4226cab8749376af5298599e0d5a1de175d9ba462608390d7d593R135

Reply:
The same to `PartitioningCollection`. It looks cleaner if the caller flattens `PartitioningCollection`, and then uses `multiTransform` to expand it.

Reply:
Ok, I will fix this and other comments today.

Reply:
ok, fixed in https://github.com/apache/spark/pull/37525/commits/a7955b58cb2cd0fe9325ca241453aca106f55af4